File: sdk/java/picodroid/view/Gravity.java
// SPDX-License-Identifier: GPL-3.0-only
package picodroid.view;

/**
 * Standard gravity constants for placing a view within a larger container. Mirrors {@code
 * android.view.Gravity} — every constant keeps Android's exact bit value, so ported code and
 * developer muscle memory ({@code Gravity.CENTER}, {@code Gravity.END}) carry over verbatim.
 *
 * <p>Pass to {@link picodroid.widget.LinearLayout#setGravity(int)} to align children along the
 * layout's main axis. The framework decodes the Android bit layout natively (LEFT/TOP → flex
 * start, RIGHT/BOTTOM → flex end, CENTER bits → flex center); {@link #START}/{@link #END} carry
 * the same axis bits as LEFT/RIGHT, so they resolve identically on this LTR-only platform.
 */
public final class Gravity {
  private Gravity() {}

  /** Constant indicating that no gravity has been set. */
  public static final int NO_GRAVITY = 0x0000;

  /** Push object to the top of its container, not changing its size. */
  public static final int TOP = 0x30;

  /** Push object to the bottom of its container, not changing its size. */
  public static final int BOTTOM = 0x50;

  /** Push object to the left of its container, not changing its size. */
  public static final int LEFT = 0x03;

  /** Push object to the right of its container, not changing its size. */
  public static final int RIGHT = 0x05;

  /** Place object in the vertical center of its container, not changing its size. */
  public static final int CENTER_VERTICAL = 0x10;

  /** Place object in the horizontal center of its container, not changing its size. */
  public static final int CENTER_HORIZONTAL = 0x01;

  /** Place the object in the center of its container in both axes, not changing its size. */
  public static final int CENTER = CENTER_VERTICAL | CENTER_HORIZONTAL;

  /** Grow the vertical size of the object if needed so it completely fills its container. */
  public static final int FILL_VERTICAL = 0x70;

  /** Grow the horizontal size of the object if needed so it completely fills its container. */
  public static final int FILL_HORIZONTAL = 0x07;

  /** Grow the object in both axes if needed so it completely fills its container. */
  public static final int FILL = FILL_VERTICAL | FILL_HORIZONTAL;

  /**
   * Push object to the beginning of its container. Matches Android's relative-direction value;
   * picodroid is LTR-only, so this behaves exactly like {@link #LEFT}.
   */
  public static final int START = 0x00800003;

  /**
   * Push object to the end of its container. Matches Android's relative-direction value;
   * picodroid is LTR-only, so this behaves exactly like {@link #RIGHT}.
   */
  public static final int END = 0x00800005;
}

File: crates/picodroid-core/src/graphics/lvgl/widgets/gravity.rs
// SPDX-License-Identifier: GPL-3.0-only
//! `LinearLayout.setGravity` as LVGL sees it: an Android gravity bitmask decoded into the flex
//! alignment of the layout's main and cross axes. Pure, so it is unit-tested through the shim in
//! `lib.rs` although the graphics tree is `cfg(not(test))`; `linear_layout.rs` applies the result.

use crate::lvgl_ffi::*;

/// Android packs a gravity into two three-bit fields: horizontal in bits 0-2, vertical in bits
/// 4-6. Each field holds `AXIS_SPECIFIED` (1), `AXIS_PULL_BEFORE` (2) and `AXIS_PULL_AFTER` (4),
/// so `LEFT` is `0x03` and `RIGHT` is `0x05` — the two ends of an axis share the SPECIFIED bit,
/// and only the pull bits tell them apart. A mask test against `LEFT` therefore matches `RIGHT`
/// too, and one against `TOP` matches `BOTTOM`; the pulls have to be read on their own.
const AXIS_SPECIFIED: i32 = 0x1;
const AXIS_PULL_BEFORE: i32 = 0x2;
const AXIS_PULL_AFTER: i32 = 0x4;

/// The three-bit field width, and the shift that reaches the vertical one.
const AXIS_MASK: i32 = 0x7;
const VERTICAL_SHIFT: i32 = 4;

/// Where one axis of a gravity places its children, or `None` when that axis says nothing —
/// `Gravity.RIGHT` alone leaves the vertical axis unspecified.
///
/// `FILL` sets both pull bits. Android stretches the child to the axis; LVGL's flex alignment
/// cannot, so it reads as the start of the axis, which is where a stretched child begins.
fn axis_align(field: i32) -> Option<lv_flex_align_t> {
    if field & AXIS_PULL_BEFORE != 0 {
        Some(LV_FLEX_ALIGN_START)
    } else if field & AXIS_PULL_AFTER != 0 {
        Some(LV_FLEX_ALIGN_END)
    } else if field & AXIS_SPECIFIED != 0 {
        // SPECIFIED with neither pull is CENTER_HORIZONTAL / CENTER_VERTICAL.
        Some(LV_FLEX_ALIGN_CENTER)
    } else {
        None
    }
}

/// The `(main, cross)` flex alignment for a gravity on a layout whose flow is a column when
/// `vertical_flow`, a row otherwise. Which Android axis is the main one follows the flow: the
/// vertical bits drive a column, the horizontal bits a row.
///
/// An axis the gravity leaves out keeps the default the layout was created with — the start of
/// the main axis, as an Android `LinearLayout` defaults to `START | TOP`, and the centre of the
/// cross axis, which is a picodroid divergence that predates per-axis routing and that no
/// caller has had a way to ask for anything else.
///
/// `Gravity.START` and `Gravity.END` carry a relative-direction bit above the two fields; masking
/// each field drops it, so they land on `LEFT` and `RIGHT` as they do in a left-to-right locale.
pub(crate) fn flex_align(gravity: i32, vertical_flow: bool) -> (lv_flex_align_t, lv_flex_align_t) {
    let horizontal = gravity & AXIS_MASK;
    let vertical = (gravity >> VERTICAL_SHIFT) & AXIS_MASK;
    let (main, cross) = if vertical_flow {
        (vertical, horizontal)
    } else {
        (horizontal, vertical)
    };
    (
        axis_align(main).unwrap_or(LV_FLEX_ALIGN_START),
        axis_align(cross).unwrap_or(LV_FLEX_ALIGN_CENTER),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    // The Gravity constants, spelled as sdk/java/picodroid/view/Gravity.java has them.
    const TOP: i32 = 0x30;
    const BOTTOM: i32 = 0x50;
    const LEFT: i32 = 0x03;
    const RIGHT: i32 = 0x05;
    const CENTER_VERTICAL: i32 = 0x10;
    const CENTER_HORIZONTAL: i32 = 0x01;
    const CENTER: i32 = CENTER_VERTICAL | CENTER_HORIZONTAL;
    const FILL_VERTICAL: i32 = 0x70;
    const START: i32 = 0x0080_0003;
    const END: i32 = 0x0080_0005;
    const NO_GRAVITY: i32 = 0;

    const ROW: bool = false;
    const COLUMN: bool = true;

    #[test]
    fn the_far_end_of_an_axis_is_not_read_as_the_near_one() {
        // The bug this function exists to rule out: RIGHT shares the SPECIFIED
        // bit with LEFT, and BOTTOM with TOP, so a mask test reads both ends as
        // the start and no layout can ever right-align.
        assert_eq!(flex_align(RIGHT, ROW).0, LV_FLEX_ALIGN_END);
        assert_eq!(flex_align(BOTTOM, COLUMN).0, LV_FLEX_ALIGN_END);
        assert_eq!(flex_align(LEFT, ROW).0, LV_FLEX_ALIGN_START);
        assert_eq!(flex_align(TOP, COLUMN).0, LV_FLEX_ALIGN_START);
    }

    #[test]
    fn the_main_axis_follows_the_flow() {
        // A vertical gravity says nothing about a row's main axis, and a
        // horizontal one nothing about a column's.
        assert_eq!(flex_align(BOTTOM, ROW).0, LV_FLEX_ALIGN_START);
        assert_eq!(flex_align(RIGHT, COLUMN).0, LV_FLEX_ALIGN_START);
        // It does reach the cross axis, which is the other half of the pair.
        assert_eq!(flex_align(BOTTOM, ROW).1, LV_FLEX_ALIGN_END);
        assert_eq!(flex_align(RIGHT, COLUMN).1, LV_FLEX_ALIGN_END);
    }

    #[test]
    fn centre_reaches_both_axes() {
        assert_eq!(
            flex_align(CENTER, COLUMN),
            (LV_FLEX_ALIGN_CENTER, LV_FLEX_ALIGN_CENTER)
        );
        assert_eq!(
            flex_align(CENTER, ROW),
            (LV_FLEX_ALIGN_CENTER, LV_FLEX_ALIGN_CENTER)
        );
        assert_eq!(flex_align(CENTER_VERTICAL, COLUMN).0, LV_FLEX_ALIGN_CENTER);
        assert_eq!(flex_align(CENTER_HORIZONTAL, ROW).0, LV_FLEX_ALIGN_CENTER);
    }

    #[test]
    fn an_unspecified_axis_keeps_the_layouts_default() {
        assert_eq!(
            flex_align(NO_GRAVITY, COLUMN),
            (LV_FLEX_ALIGN_START, LV_FLEX_ALIGN_CENTER)
        );
        // Only the main axis named: the cross keeps centring, as before.
        assert_eq!(flex_align(BOTTOM, COLUMN).1, LV_FLEX_ALIGN_CENTER);
    }

    #[test]
    fn relative_start_and_end_land_on_left_and_right() {
        assert_eq!(flex_align(START, ROW).0, LV_FLEX_ALIGN_START);
        assert_eq!(flex_align(END, ROW).0, LV_FLEX_ALIGN_END);
    }

    #[test]
    fn fill_degrades_to_the_start_of_the_axis() {
        assert_eq!(flex_align(FILL_VERTICAL, COLUMN).0, LV_FLEX_ALIGN_START);
    }

    #[test]
    fn a_gravity_naming_both_axes_places_both() {
        assert_eq!(
            flex_align(BOTTOM | RIGHT, COLUMN),
            (LV_FLEX_ALIGN_END, LV_FLEX_ALIGN_END)
        );
        assert_eq!(
            flex_align(TOP | RIGHT, ROW),
            (LV_FLEX_ALIGN_END, LV_FLEX_ALIGN_START)
        );
    }
}
